src/libutil/fmath_test.cpp

```diff
@@ -144,6 +145,50 @@ test_int_helpers ()
 
 
 
+void
+test_math_functions ()
+{
+    std::cout << "Testing math functions\n";
+    Benchmarker bench;
+
+    OIIO_CHECK_EQUAL (ifloor(0.0f), 0);
+    OIIO_CHECK_EQUAL (ifloor(-0.999f), -1);
+    OIIO_CHECK_EQUAL (ifloor(-1.0f), -1);
+    OIIO_CHECK_EQUAL (ifloor(-1.001f), -2);
+    OIIO_CHECK_EQUAL (ifloor(0.999f), 0);
+    OIIO_CHECK_EQUAL (ifloor(1.0f), 1);
+    OIIO_CHECK_EQUAL (ifloor(1.001f), 1);
+    float fval = 1.1; clobber(fval);
+    bench ("ifloor", [&](){ return DoNotOptimize(ifloor(fval)); });
```

Comment:
Which architecture did  you benchmark on? Because from what I looked up - `roundss` is a single cycle op on modern chips. It seems funny the extra comparison and all that would ever win.

Reply:
In fact I seem to recall that in other places where we were using this trick, switching to plain `floorf` got faster as soon as we compiled with `sse4.1+`.

Reply:
I will double check today by rebuilding/running for every SIMD architecture (on the work machines) and post the results.

I think the floor (which is roundss/roundps underneath) is not the expensive part, it's the cast to int that really hurts.

Reply:
Right - but both functions cast to int ...

Reply:
On the work Linux machines:  (all numbers are Mvalues/second, bigger is better)

USE_SIMD=sse2:

    float ifloor:  543
    float fast_ifloor:   1348
    vfloat4 ifloor:   501
    vfloat4 fast_ifloor:   6400
    vfloat8 ifloor:   501
    vfloat8 fast_ifloor:   6394

USE_SIMD=sse4.1

    float ifloor:  1066
    float fast_ifloor: 1348
    vfloat4 ifloor:  4264
    vfloat4 fast_ifloor:  6400
    vfloat8 ifloor:  4259
    vfloat8 fast_ifloor:  6390

USE_SIMD=avx2

    float ifloor: 1066
    float fast_ifloor:  1348
    vfloat4 ifloor:  4264
    vfloat4 fast_ifloor:  6400
    vfloat8 ifloor:  8510
    vfloat8 fast_ifloor:  12800

Reply:
godbolt.org:

```
#include <cmath>

int ifloor (float x)
{
    return (int)floorf(x);
}

int fast_ifloor (float x)
{
    return (int) x - (x < 0.0f ? 1 : 0);
}
```

For sse2, you get:

```
floorf@plt:
 jmpq   *0x200bea(%rip)        # 601020 <_GLOBAL_OFFSET_TABLE_+0x20>
 pushq  $0x1
 jmpq   400410 <_init+0x20>
.plt.got:
 jmpq   *0x200bb2(%rip)        # 600ff8 <_DYNAMIC+0x200>
 xchg   %ax,%ax
ifloor(float):
 push   %rax
 callq  400430 <floorf@plt>
 cvttss2si %xmm0,%eax
 pop    %rcx
 retq   
 nopl   0x0(%rax)
fast_ifloor(float):
 cvttss2si %xmm0,%eax
 xorps  %xmm1,%xmm1
 xor    %ecx,%ecx
 ucomiss %xmm0,%xmm1
 seta   %cl
 sub    %ecx,%eax
 retq   
 data16 data16 data16 data16 nopw %cs:0x0(%rax,%rax,1)
main:
 xor    %eax,%eax
 retq   
 nopw   %cs:0x0(%rax,%rax,1)
 nopl   (%rax)
```

For sse4.2:
```
.plt.got:
 jmpq   *0x200c12(%rip)        # 600ff8 <_DYNAMIC+0x200>
 xchg   %ax,%ax
ifloor(float):
 roundss $0x9,%xmm0,%xmm0
 cvttss2si %xmm0,%eax
 retq   
 nopl   0x0(%rax,%rax,1)
fast_ifloor(float):
 cvttss2si %xmm0,%eax
 xorps  %xmm1,%xmm1
 xor    %ecx,%ecx
 ucomiss %xmm0,%xmm1
 seta   %cl
 sub    %ecx,%eax
 retq   
 data16 data16 data16 data16 nopw %cs:0x0(%rax,%rax,1)
main:
 xor    %eax,%eax
 retq   
 nopw   %cs:0x0(%rax,%rax,1)
 nopl   (%rax)
```

And for avx2:
```
.plt.got:
 jmpq   *0x200c12(%rip)        # 600ff8 <_DYNAMIC+0x200>
 xchg   %ax,%ax
ifloor(float):
 vroundss $0x9,%xmm0,%xmm0,%xmm0
 vcvttss2si %xmm0,%eax
 retq   
 nopl   0x0(%rax,%rax,1)
fast_ifloor(float):
 vcvttss2si %xmm0,%eax
 vxorps %xmm1,%xmm1,%xmm1
 xor    %ecx,%ecx
 vucomiss %xmm0,%xmm1
 seta   %cl
 sub    %ecx,%eax
 retq   
 data16 data16 nopw %cs:0x0(%rax,%rax,1)
main:
 xor    %eax,%eax
 retq   
 nopw   %cs:0x0(%rax,%rax,1)
 nopl   (%rax)
```

So SSE2 is super slow because it's making a function call into libm, whereas sse4.2 is using roundss.

But for sse4.2 and avx2, I don't understand why fast_ifloor is consistently faster than the two instructions for ifloor. Is it possible ifloor's vroundss followed by a vcvtss2si produces some kind of stall that's worse than fast_ifloor's vcvtss2i followed by several integer ops?